java-checks/src/main/java/org/sonar/java/checks/BoxedBooleanExpressionsCheck.java

```diff
@@ -0,0 +1,119 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import java.util.Arrays;
+import java.util.List;
+
+import javax.annotation.CheckForNull;
+
+import org.sonar.check.Rule;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.tree.BinaryExpressionTree;
+import org.sonar.plugins.java.api.tree.ConditionalExpressionTree;
+import org.sonar.plugins.java.api.tree.DoWhileStatementTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.ForStatementTree;
+import org.sonar.plugins.java.api.tree.IfStatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.WhileStatementTree;
+import org.sonar.plugins.java.api.tree.Tree.Kind;
+
+@Rule(key = "S5411")
+public class BoxedBooleanExpressionsCheck extends IssuableSubscriptionVisitor {
+
+  private static final String BOOLEAN = "java.lang.Boolean";
+
+  @Override
+  public List<Kind> nodesToVisit() {
+    return Arrays.asList(Kind.FOR_STATEMENT, Kind.WHILE_STATEMENT, Kind.DO_STATEMENT, Kind.IF_STATEMENT, Kind.EQUAL_TO,
+        Kind.NOT_EQUAL_TO, Kind.CONDITIONAL_EXPRESSION, Kind.CONDITIONAL_AND, Kind.CONDITIONAL_OR);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+    ExpressionTree expression;
+    switch (tree.kind()) {
+      case FOR_STATEMENT:
+        expression = ((ForStatementTree) tree).condition();
+        break;
+      case WHILE_STATEMENT:
+        expression = ((WhileStatementTree) tree).condition();
+        break;
+      case DO_STATEMENT:
+        expression = ((DoWhileStatementTree) tree).condition();
+        break;
+      case IF_STATEMENT:
+        expression = ((IfStatementTree) tree).condition();
+        break;
+      case CONDITIONAL_EXPRESSION:
+        expression = ((ConditionalExpressionTree) tree).condition();
+        break;
+      case EQUAL_TO:
+      case NOT_EQUAL_TO:
+      case CONDITIONAL_AND:
+      case CONDITIONAL_OR:
+        expression = (BinaryExpressionTree) tree;
+        break;
+      default:
+        expression = null;
+    }
+    if (expression != null) {
+      ExpressionTree boxed = findBoxedBoolean(expression);
+      if (boxed != null) {
+        reportIssue(boxed, "Use the primitive boolean expression here.");
+      }
+    }
+  }
+
+  @CheckForNull
+  private static ExpressionTree findBoxedBoolean(ExpressionTree tree) {
+    if (isBooleanIdentifier(tree)) {
+      return tree;
+    }
+    if (tree instanceof BinaryExpressionTree) {
+      BinaryExpressionTree expression = (BinaryExpressionTree) tree;
```

Comment:
In case of binary expression we should just check that either lhs or rhs is `java.lang.Boolean` and `boolean` respectively, no need to check for null checks.